### PR TITLE
[APMSVLS-485] fix(traces): bound trace stats cardinality and report collapsed keys

### PR DESCRIPTION
**Pull request by lucaspimentel**
Stacked PRs:
- DataDog/serverless-components#141
- #1332
- #1338 👈🏽 **This PR**
- #1336

## Overview

Adopts libdatadog's default trace-stats cardinality limits, and reports it when they collapse
keys. Split out of #1332 so the customer-visible half of that bump gets reviewed on its own
rather than buried in a `chore(deps)` PR.

### 1. `fix(traces): bound trace stats cardinality`

Pass `Some(CardinalityLimitConfig::default())` and take libdatadog's defaults (`None` is equivalent — libdatadog resolves it with `unwrap_or_default()` — but passing the defaults explicitly makes the limits the collapse warnings quote provably the ones in force, instead of a second copy that could drift) — 7000 whole-key, 1024
resource, 512 http endpoint, 512 peer tags, 100 additional tags. #1332 pins all five to
effectively-unbounded values to preserve pre-bump behavior, which also opts out of the four *new*
per-field protections: a function whose span resources embed request ids or path parameters
yields one aggregation group per distinct resource per 10s bucket, each carrying two ddsketches,
so concentrator memory and the `/v0.6/stats` payload grow without bound inside a memory-capped
Lambda. Defaults also restore parity with SCL's `datadog-trace-agent` (same libdatadog rev) and
the Go agent, both of which keep them.

**User-visible:** for functions with high-cardinality span resources, resources past the 1,024th
in a 10-second bucket are now reported under `tracer_blocked_value` instead of individually. That
is the intended trade for bounded memory.

This also removes the four unbounded per-field tracking `HashSet<u64>`s that #1332's pinning
necessarily leaves behind (bounding them *is* collapsing, so no pinning avoids both), which is
why these two PRs should not be split across a release.

### 2. `feat(traces): report collapsed stats keys`

Collapse is otherwise silent — keys are rewritten to `tracer_blocked_value` and the affected
stats stop being attributable, with nothing in the logs. Warn once per sandbox per signal.

**Two signals are needed, because they fail in opposite directions:**

| signal | source | carries a count? |
|---|---|---|
| whole-key overflow | `FlushResult.collapsed_spans` | yes |
| per-field collapse | scan the flushed payload for the sentinel | no |

`collapsed_spans` alone would be silent for the canonical Lambda failure. Per-field limits are
applied *before* the whole-key limit (`aggregation.rs:589` runs ahead of the check at `:603`), so
one dimension exploding — request ids in resource names — collapses resources at 1,024 first,
which shrinks the distinct whole-key space and can stop `collapsed_spans` ever leaving 0. The
counter bottlecap can read is the one least likely to fire on the case that matters.

The payload scan covers it: per-field collapse rewrites *only* the field that overflowed, while
the whole-key overflow entry has *every* field set to the sentinel, so `service` — never rewritten
per-field — identifies and skips that entry. Otherwise one whole-key overflow would masquerade as
all four fields collapsing at once.

**Why not libdatadog's per-field counters.** `StatsBucket::collapsed_fields_metrics()` is
reachable and `Copy`, but its per-combination counts have no public accessor without the
`dogstatsd`/`telemetry` features, which bottlecap does not enable (`libdd-trace-stats` is declared
`default-features = false`, only `https`/`fips`). Once upstream exposes them, this can report
*how many* keys collapsed rather than only which fields. Tracked as a follow-up libdatadog PR we will author (not a Jira ticket); not a blocker for this stack.

**Why no environment variable is named.** The limits are not customer-tunable here, and both
candidates mislead: libdatadog's own message blames `DD_TRACE_STATS_CARDINALITY_LIMIT`, which
bottlecap does not read at all, and `DD_TRACE_STATS_ADDITIONAL_TAGS_CARDINALITY_LIMIT` governs
only `additional_tags`. Reducing application-side cardinality is the only real remediation, so
that is what the warnings recommend. No per-flush `debug!` either — libdatadog already emits one
for whole-key overflow at `span_concentrator/mod.rs:442`, and it would be a near-duplicate.

`TRACER_BLOCKED_VALUE` is hand-copied: upstream declares it `pub`, but `mod aggregation` is
private and it is not re-exported, so it cannot be imported. A one-line upstream `pub use` would
remove the copy; also tracked as a follow-up libdatadog PR.

### 3. Review follow-ups

Three small commits on top, all in `stats_concentrator_service.rs`:

- `style(traces): remove em dashes from stats collapse warnings` — the per-field warning is
  customer-facing, so it is rewritten with ordinary punctuation, along with the doc comments
  around it.
- `fix(traces): derive bucket duration in collapse warning` — that warning hardcoded "in a 10s
  bucket", which would have gone stale if the aggregation bucket duration ever changed; it now
  reads the value from `BUCKET_DURATION_NS`.
- `perf(traces): stop scanning stats payloads once collapse is reported` — each signal warns at
  most once per sandbox, so once all of them have fired `report_collapse` can only recompute a
  result it throws away. It now returns early: a bucket can hold thousands of aggregation entries
  and flushes are frequent under continuous flushing. No new test, the behavior is unobservable
  through the public API and `test_collapse_warns_once_per_signal` already pins the once-per-signal
  guarantee it relies on.

## Testing

- `cargo test` — clean
- `cargo clippy --all-targets -- -D warnings` — clean
- `cargo fmt --check` — clean

New tests, deliberately kept distinct because the obvious ones assert the wrong thing:

- **`test_observe_collapsed_fields`** — each field detected independently; the whole-key overflow
  entry is skipped rather than counted as all four collapsing; both tag lists covered, which
  encode a valueless sentinel differently (`peer_tags` as the bare key,
  `additional_metric_tags` with a trailing colon); collapse in any bucket of the flush counts.
- **`test_resource_collapse_observed_without_whole_key_overflow`** — the regression test for the
  trap above. Exceeds `resource_limit` (1,024) while staying under `whole_key_limit` (7,000), and
  asserts the collapse *is* observed while asserting no whole-key overflow entry exists.
- **`test_collapse_warns_once_per_signal`** — each signal warns at most once, and the two are
  independent, so whole-key overflow does not suppress a later per-field collapse. Also the only
  coverage of the whole-key branch: `new()` always builds the concentrator with libdatadog's
  defaults, so driving keys past 7,000 without first tripping 1,024 is not reachable through the
  public API. Asserts on service state, not log output.
- **`test_cardinality_limit_applied`** — exceeds `resource_limit` by two distinct resources
  with a shared timestamp, so every span lands in the same bucket. Asserts the admitted resource
  groups plus exactly one sentinel overflow group, and verifies that the overflow and total hit
  counts retain every inserted span. It says nothing about `collapsed_spans`; it is not a
  substitute for the two tests above.

### CODE REVIEW

**Comment by lucaspimentel** [comment]
@codex review

File: bottlecap/src/traces/stats_concentrator_service.rs [hunk 1]
```diff
@@ -74,6 +82,94 @@ const DEFAULT_PEER_TAG_KEYS: &[&str] = &[
...
+/// overflow entry has *every* field set to the sentinel. `service` is never rewritten by per-field
+/// collapse, so it identifies that overflow entry and lets us skip it — otherwise a single
+/// whole-key overflow would masquerade as all four fields collapsing at once.
+fn observe_collapsed_fields(buckets: &[pb::ClientStatsBucket]) -> CollapsedFields {
+    let mut observed = CollapsedFields::default();
+    for stats in buckets.iter().flat_map(|bucket| &bucket.stats) {
+        if stats.service == TRACER_BLOCKED_VALUE {
+            continue;
+        }
+        if stats.resource == TRACER_BLOCKED_VALUE {
```
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply bottlecap/src/traces/stats_concentrator_service.rs:159]: Skipping this. The sentinel scan is deliberate, not an oversight: `StatsBucket::collapsed_fields_metrics()` has no public accessor without the `dogstatsd`/`telemetry` features at this libdatadog pin, so scanning the payload for `tracer_blocked_value` is the only signal bottlecap can get for per-field collapse today. A real resource/service/endpoint literally named `tracer_blocked_value` colliding with the sentinel is an accepted edge case (same approach libdatadog's own test suite uses). Exact per-field counts without inferring from values is tracked as a separate upstream libdatadog follow-up.

🤖

[the remark this answers is not in the document]
Comment (lucaspimentel) [reply bottlecap/src/traces/stats_concentrator_service.rs:159]: Skipping this. Duplicate of https://github.com/DataDog/datadog-lambda-extension/pull/1338#discussion_r3855036013

File: bottlecap/src/traces/stats_concentrator_service.rs [hunk 2]
```diff
@@ -283,6 +388,64 @@ impl StatsConcentratorService {
...
+    ///
+    /// Two independent signals, because they fail in opposite directions:
+    ///
+    /// - **whole-key** overflow, from `collapsed_spans`, which carries a span count;
+    /// - **per-field** collapse, from scanning the payload, which does not.
+    ///
+    /// Both are needed. Per-field limits are applied *before* the whole-key limit
+    /// (`aggregation.rs:589` runs ahead of the check at `:603`), so a single-dimension explosion —
+    /// request ids in resource names, the canonical Lambda failure — collapses resources first,
+    /// which shrinks the distinct whole-key space and can stop `collapsed_spans` ever leaving 0.
```
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply bottlecap/src/traces/stats_concentrator_service.rs]: Fixed by condensing the code comment (dropped line-number references and history).

**Review by litianningdatadog — APPROVED** [review]
LGTM
